User defined kill signal value

v-p-b opened the pull request:
This adds support for the AFL_KILL_SIGNAL environment variable discussed here:

https://github.com/intel/kernel-fuzzer-for-xen-project/pull/24#issuecomment-755271472

I need some guidance regarding:

* The additions made to support hang detection when running with KF/x, see:
https://github.com/AFLplusplus/AFLplusplus/compare/stable...v-p-b:env_kill?expand=1#diff-2d26035103ef709cb5d8fe28c0ed22a9554333e1dedf3f08dff919038488ec11R1146
https://github.com/intel/kernel-fuzzer-for-xen-project/pull/24#issuecomment-755711776
* Whether the environment variable is handled the right way (`getenv()` vs. `get_afl_env()`, `fsrv` struct extension...)

I will merge the commits from bleeding edge (this is currently forked from 3.0c), and squash my modifications if requested.

v-p-b committed:
Adding AFL_KILL_SIGNAL environment variable

Controlling the kill signal used to end forked processes.

v-p-b committed:
Removed unnecessary afl state struct member

v-p-b committed:
Using getenv() instead of afl_get_env() in at_exit()

vanhauser-thc commented:
Hi! Thanks for the PR.
we only accept PRs to dev, never to stable, so that stable never has breaking code. I therefore switched the PR to point to dev.
I will now review the changes and make comments if necessary.

vanhauser-thc commented on src/afl-forkserver.c:
```diff
@@ -95,6 +95,10 @@ void afl_fsrv_init(afl_forkserver_t *fsrv) {
+  fsrv->kill_signal = SIGKILL;
+  if (get_afl_env("AFL_KILL_SIGNAL")){
+    fsrv->kill_signal = atoi(get_afl_env("AFL_KILL_SIGNAL"));
+  }
```
this needs error handling in the case that e.g. ` AFL_KILL_SIGNAL=foobar`  or an invalid signal number. also people are used to use the signal name instead of the signal number.

vanhauser-thc replied:
I recommend to reset the env var with an ensured correct value here so the same code is not needed in `at_exit()`

domenukk replied:
Using a number was my initial idea, is there any way to get a mapping of string -> id? Personally, I'm rather used to doing kill 9

domenukk replied:
Answering my own question, this is how kill does it:
https://github.com/coreutils/coreutils/blob/00ea4bacf6063ccc125209d5186f8f2382c6f0d4/src/operand2sig.c#L36
But I would argue users can look up the table instead

vanhauser-thc replied:
ok, however it has still to be checked that the env var is a valid number and FATAL if not.

tklengyel replied:
I would be fine with using just a number.

domenukk replied:
We try not to parse env variables here, but instead set the kill signal when parsing, before starting the forkserver.

vanhauser-thc commented on src/afl-fuzz-state.c:
```diff
@@ -507,8 +507,8 @@ void afl_states_stop(void) {
 
-    if (el->fsrv.child_pid > 0) kill(el->fsrv.child_pid, SIGKILL);
-    if (el->fsrv.fsrv_pid > 0) kill(el->fsrv.fsrv_pid, SIGKILL);
+    if (el->fsrv.child_pid > 0) kill(el->fsrv.child_pid, el->fsrv.kill_signal);
```
I think it needs a second variable `AFL_DO_NOT_KILL_CHILD` or something like that? I think I remember @tklengyel writing that the child may not be killed for Xen mode.

domenukk replied:
This was resolved on xen mode side

tklengyel replied:
Yes, this is not an issue anymore, we report the forkserver's pid as the child's pid and that just works. The forkserver receives the signal twice here but that doesn't matter.

v-p-b committed:
code formatting

v-p-b committed:
Merge branch 'dev' of https://github.com/AFLplusplus/AFLplusplus into env_kill

domenukk commented:
WRT adding new env variables, it's a bit more involved, see https://github.com/AFLplusplus/AFLplusplus/pull/522/files

(Surprising, given how many env vars we have :P)

/edit: the linked pr is still missing env var parsing for showmap and tmin...

v-p-b committed:
Checking validity of AFL_KILL_SIGNAL env variable

This commit also sets a valid value in the environment to avoid
duplicating code in at_exit(). Changing data type of
fsrv->kill_signal to u8 to match last_kill_signal.

v-p-b committed:
Adding afl_kill_signal to AFL (environment) state

This commit simply introduces a struct member for future use. The
env variable is not used from the afl struct but from fsrv, where
its validity is checked, resulting in a FATAL in case of errors.

v-p-b commented:
I pushed two commits with long messages to describe my intentions - I hope they resolve at least some of the issues you raised! 

Tests run fine on my Ubuntu 20.04 - can someone take a look at what can be the problem with the CI?

domenukk commented on src/afl-forkserver.c:
```diff
@@ -1137,6 +1162,15 @@ fsrv_run_result_t afl_fsrv_run_target(afl_forkserver_t *fsrv, u32 timeout,
 
+  /* TODO We use SIGTERM here as an indicator of Xen mode,
+     although it's not equivalent! */
+  if (fsrv->kill_signal == SIGTERM && !*stop_soon_p &&
```
Pretty sure we can just always do the `last_run_timed_out` check here, and drop it below, any reason not to?
Is there a race condition here the recent code could catch?

Also, we can wrap it in a `unlikely`

domenukk replied:
Ah I guess it is one (usually) unneeded jump? Well we don't save that jump by adding another condition, either :)

domenukk reviewed (approved):
I think getting rid of the wifsignaled check before timeouts completely is fine, what do y'all think?
Apart from that, thanks a lot @v-p-b I can take it from here :)

domenukk commented:
Cleaned it up in https://github.com/AFLplusplus/AFLplusplus/commit/a06b25538fd7b6eef6755094aa4678c2cb5333fd maybe somebody can take a look
